### PR TITLE
HIVE-27000: Improve the modularity of the *ColumnStatsMerger classes

### PR DESCRIPTION
**Pull request by asolimando**
### What changes were proposed in this pull request?

Improving the modularity of the *ColumnStatsMerger classes by introducing a common parent class and moving there all the shared code that is not depending on the data type.

### Why are the changes needed?

Those classes host some complex merging logic that is copy-pasted into several classes, a problem that is bound to worsen as we add new (and more complex) statistics.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing and introduced unit-tests relevant for the change: 
```mvn test -Dtest.groups=org.apache.hadoop.hive.metastore.annotation.MetastoreUnitTest -Dtest='*ColumnStats*Test.java' -pl standalone-metastore/metastore-server```

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/merge/DateColumnStatsMerger.java [hunk 1]
```diff
@@ -43,64 +46,57 @@ public void merge(ColumnStatisticsObj aggregateColStats, ColumnStatisticsObj new
...
-      }
-      LOG.debug("Use bitvector to merge column {}'s ndvs of {} and {} to be {}", aggregateColStats.getColName(),
-          aggregateData.getNumDVs(), newData.getNumDVs(), ndv);
-      aggregateData.setNumDVs(ndv);
+    Date lowValue = mergeLowValue(getLowValue(aggregateData), getLowValue(newData));
+    if (lowValue != null) {
+      aggregateData.setLowValue(lowValue);
+    }
+    Date highValue = mergeHighValue(getHighValue(aggregateData), getHighValue(newData));
+    if (highValue != null) {
```
Comment (akshat0395) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/merge/DateColumnStatsMerger.java:54]: This code block seems to be duplicate for multiple StatsMerger classes, can we move this to a shared class to avoid duplication.

**Review by akshat0395 — CHANGES_REQUESTED** [review]

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/merge/DateColumnStatsMerger.java (hunk 1 above)
Comment (asolimando) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/merge/DateColumnStatsMerger.java:54]: Thanks @akshat0395 for your comment, you are right, but unfortunately the problem lies in the `*ColumnStatsDataInspector` objects, which inherits from the corresponding `*ColumnStatsData`, which don't have a common ancestor.

For this reason it's hard to write a generic method, we would need to rely on generic for the   data type (say `Double`) and anyway a big switch for handling the *ColumnStatsDataInspector objects and casting, it would rapidly defeat the purpose.

This lack of class hierarchy for column statistics comes directly from the thrift interface objects for the column statistics in the metastore (see [here](https://github.com/apache/hive/blob/master/standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift)).

HIVE-26313 could be a chance to overcome that, and define a better hierarchy for those classes, since we won't anyway be having a 1-1 mapping from a column in the metastore for column statistics and a Java class, but we will have to parse a (probably Json) blob and do the "mapping" ourselves.

In summary, I'd rather leave this for when we will have a better class hierarchy for those objects (hopefully with HIVE-26313). WDYT?

Comment (akshat0395) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/columnstats/merge/DateColumnStatsMerger.java:54]: Thanks @asolimando for the explanation, if there is a plan improve the class hierarchy then it make sense to tackle this then as well.

**Comment by akshat0395** [comment]
LGTM, +1

**Review by akshat0395 — APPROVED** [review]

**Comment by asolimando** [comment]
@ayushtkn, if you have spare cycles maybe you can help get this in since it's been approved already? Thanks!

**Review by ayushtkn — APPROVED** [review]
LGTM.
Will give a day before committing in case anyone else has comments

**Comment by ayushtkn** [comment]
And I messed up while committing :-( 
Forgot to mention Akshat's name while committing, Have reverted this and committed it again with the correct message. I don't have anything more than sorry here, will be more careful next time

**Comment by asolimando** [comment]
> And I messed up while committing :-( Forgot to mention Akshat's name while committing, Have reverted this and committed it again with the correct message. I don't have anything more than sorry here, will be more careful next time

We never make mistakes only if we do nothing at all, you are very very active so it happens. On that note, thanks a lot for always helping out with reviews and merges, much appreciated Ayush!
